x-pack/plugin/snapshot-based-recoveries/src/internalClusterTest/java/org/elasticsearch/xpack/snapshotbasedrecoveries/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -1200,6 +1255,142 @@ public void testRecoveryRetryKeepsTheGrantedSnapshotFileDownloadPermit() throws
+            .filter(nodeName -> nodeName.equals(primaryNodeName) == false)
+            .filter(nodeName -> nodeName.equals(replicaNodeName) == false)
+            .findFirst()
+            .orElseThrow();
+
+        MockTransportService sourceMockTransportService = (MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            replicaNodeName
+        );
+
+        MockTransportService targetMockTransportService = (MockTransportService) internalCluster().getInstance(
+            TransportService.class,
+            newReplicaNodeName
+        );
+
+        final CountDownLatch firstDownloadStartLatch = new CountDownLatch(1);
+        final CountDownLatch blockSnapshotFileDownload = new CountDownLatch(1);
+
+        final AtomicBoolean firstDataBlobRead = new AtomicBoolean();
+        FilterFsRepository.wrapReadBlobMethod((blobName, inputStream) -> {
+            if (blobName.startsWith("__") && firstDataBlobRead.compareAndSet(false, true)) {
+                return new FilterInputStream(inputStream) {
+                    @Override
+                    public int read(byte[] b, int off, int len) throws IOException {
+                        firstDownloadStartLatch.countDown();
+                        try {
+                            blockSnapshotFileDownload.await();
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                        return super.read(b, off, len);
+                    }
+                };
+            } else {
+                return inputStream;
+            }
+        });
+
+        Set<Transport.Connection> connectionRefs = Collections.synchronizedSet(new HashSet<>());
+        sourceMockTransportService.addSendBehavior(targetMockTransportService, (connection, requestId, action, request, options) -> {
+            if (action.equals(PeerRecoveryTargetService.Actions.RESTORE_FILE_FROM_SNAPSHOT)) {
+                connectionRefs.add(connection);
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+
+        assertThat(internalCluster().stopNode(primaryNodeName), is(equalTo(true)));
+
+        firstDownloadStartLatch.await();
+        // Increase the retry timeout as it takes a bit until the source node gets reconnected to the target node
+        updateSetting(INDICES_RECOVERY_INTERNAL_ACTION_RETRY_TIMEOUT_SETTING.getKey(), "1m");
+        assertThat(connectionRefs, is(not(empty())));
+        connectionRefs.forEach(Transport.Connection::close);
+
+        CountDownLatch firstFileChunkSent = new CountDownLatch(1);
+        CountDownLatch blockFileChunkDownload = new CountDownLatch(1);
+
+        targetMockTransportService.addRequestHandlingBehavior(
+            PeerRecoveryTargetService.Actions.FILE_CHUNK,
+            (handler, request, channel, task) -> {
+                firstFileChunkSent.countDown();
```

Comment:
I am not sure I see how we are sure this has not completed already before adding the request handling behavior?

Reply:
Yes, that's a bit racy. I'll move it before we close the connections. Good catch.